runtime/lua/vim/lsp/_capability.lua

```diff
@@ -69,9 +77,137 @@ function M:destroy()
   self.active[self.bufnr] = nil
 end
 
+--- Callback invoked when an LSP client attaches.
+--- Use it to initialize per-client state (empty table, new namespaces, etc.),
+--- or issue requests as needed.
+---@param client_id integer
+function M:on_attach(client_id)
+  self.client_state[client_id] = {}
+end
+
+--- Callback invoked when an LSP client detaches.
+--- Use it to clear per-client state (cached data, extmarks, etc.).
 ---@param client_id integer
 function M:on_detach(client_id)
   self.client_state[client_id] = nil
```

Comment:
would this be needed if the state is stored on the client itself, instead?

also, is it worth using a weak table here? https://www.lua.org/pil/17.html